grails-app/services/org/pih/warehouse/importer/InventoryLevelImportDataService.groovy

```diff
@@ -10,27 +10,217 @@
 package org.pih.warehouse.importer
 
 import grails.gorm.transactions.Transactional
-import org.pih.warehouse.data.DataService
+import grails.plugins.csv.CSVWriter
+import org.grails.plugins.excelimport.ExpectedPropertyType
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.inventory.InventoryLevel
+import org.pih.warehouse.inventory.InventoryStatus
+import org.pih.warehouse.product.Product
 
 @Transactional
 class InventoryLevelImportDataService implements ImportDataService {
-    DataService dataService
 
     @Override
     void validateData(ImportDataCommand command) {
-        println "validate inventory levels " + command.filename
+        log.info "validate inventory levels " + command.filename
+
         command.data.eachWithIndex { row, index ->
-            if (!dataService.validateInventoryLevel(row)) {
+            if (!validateInventoryLevel(row)) {
                 command.errors.reject("Row ${index + 1}: Failed validation")
             }
         }
     }
 
     @Override
     void importData(ImportDataCommand command) {
-        println "Import inventory levels " + command.location + " filename " + command.filename
-        command.data.eachWithIndex { row, index ->
-            dataService.importInventoryLevel(command.location, row, index)
+        command.data.eachWithIndex { Map params, Integer index ->
+            importInventoryLevel(params, index)
+        }
+    }
+
+    def validateInventoryLevel(Map params) {
+        params.each { key, value ->
+            def expectedType = InventoryLevelExcelImporter.propertyMap.get(key).expectedType
+            switch (expectedType) {
+                case ExpectedPropertyType.IntType:
+                    assert !value || value instanceof Number || value instanceof Boolean, "Value [${value}] for column [${key}] must be a Number or Boolean but was ${value?.class?.name} (" + row + ")."
+                    break
+                case ExpectedPropertyType.DateType:
+                    assert !value || value instanceof Date, "Value [${value}] for column [${key}] must be a Date but was ${value?.class?.name} (" + params + ")."
+                    break
+                default:
+                    break
+            }
+        }
+        return true
+    }
+
+    /**
+     * Import a single row from the XLS file
+     *
+     * @param location
+     * @param fileName
+     * @return
+     */
+    def importInventoryLevel(Map params, Integer index) {
+        log.info "Import inventory levels(${index}): " + params
+
+        def product = Product.findByProductCode(params.productCode as String)
```

Comment:
I think it's type of Product

Reply:
This is used to "cast" the value returned from the params map (that was populated by the excel importer). For some reason, the params map always returns an Object instead of the type we've defined in the property map (i.e. String, Integer, Date). So using "as" converts it to the expected type. I wish I didn't have to do it, but I suspect our beef is with the AbstractExcelImporter.

Reply:
Sorry, just realized this was one of the places where "def" was used. Yeah this was existing code that I forgot to change. Will push a fix later.

Reply:
> this was existing code

That's my bad, I didn't notice that

Reply:
No worries, you can/should still call it out. It's probably best to "boy scout" these kinds of refactorings if time permits.